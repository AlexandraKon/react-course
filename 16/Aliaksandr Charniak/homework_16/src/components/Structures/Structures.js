import React, { useEffect } from 'react'
import { useDispatch } from 'react-redux'

import { getCorrectedArray } from '../../utils/utils'
import { ACTION_GET_STRUCTURES_REQUESTED } from '../../redux/structures/actions'

export const Structures = (props) => {
  const { data } = props.data
  const { state } = props.location
  //const arr = data ? (data.structures ? data.structures : Array.isArray(data) ? data : [data]) : []
  const arr = getCorrectedArray(data, 'structures')
  const dispatch = useDispatch()
  useEffect(() => {
    dispatch(ACTION_GET_STRUCTURES_REQUESTED(state?.structUrl))
  }, [dispatch, state?.structUrl])
  return (
    <div>
      {arr &&
        arr.map((item) => (
          <ul key={item.id}>
            <li>Id: {item.id}</li>
            <li>Name: {item.name}</li>
            <li>Expansion: {item.expansion}</li>
            <li>Age: {item.age}</li>
            {item.cost && (
              <li>
                Cost:
                {Object.keys(item.cost).map((key, i) => {
                  return (
                    <ul key={i}>
                      <li>
                        {key}: {item.cost[key]}
                      </li>
                    </ul>
                  )
                })}
              </li>
            )}
            <li>Build time: {item.build_time}</li>
            <li>Hit Points: {item.hit_points}</li>
            <li>Line of Sight: {item.line_of_sight}</li>
            <li>Armor: {item.armor}</li>
            {item.special && (
              <li>
                Special:
                {item.special.map((it, i) => (
                  <ul key={i}>
                    <li>{it}</li>
                  </ul>
                ))}
              </li>
            )}
            <li>Reload time: {item.reload_time}</li>
            <li>Range: {item.range}</li>
            <li>Attack: {item.attack}</li>
          </ul>
        ))}
    </div>
  )
}

export default Structures
