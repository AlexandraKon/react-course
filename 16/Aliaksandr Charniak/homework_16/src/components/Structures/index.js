import { connect } from 'react-redux'

import Structures from './Structures'

const mapStateToProps = (state) => ({
  data: state.structures,
})

export default connect(mapStateToProps)(Structures)
