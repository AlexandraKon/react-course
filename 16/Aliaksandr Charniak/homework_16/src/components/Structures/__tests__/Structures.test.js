import React from 'react'
import { configure, shallow } from 'enzyme'
import Adapter from '@wojtekmaj/enzyme-adapter-react-17'
import configureMockStore from 'redux-mock-store'

import { Provider } from 'react-redux'

import Structures from '..'

configure({ adapter: new Adapter() })

const mockStore = configureMockStore()
const store = mockStore({})

describe('<Structures /> mounting behaviour', () => {
  const props = {
    data: {
      structures: [
        {
          id: 1,
          name: 'Barracks',
          expansion: 'Age of Kings',
          age: 'Dark',
          cost: {
            Wood: 175,
          },
          build_time: 50,
          hit_points: 1200,
          line_of_sight: 5,
          armor: '0/7',
          special: ['Garrison: 10 created units'],
        },
        {
          id: 2,
          name: 'Dock',
          expansion: 'Age of Kings',
          age: 'Dark',
          cost: {
            Wood: 150,
          },
          build_time: 35,
          hit_points: 1800,
          line_of_sight: 5,
          armor: '0/7',
          special: ['Garrison: 10 created units'],
        },
      ],
      units: 'https://age-of-empires-2-api.herokuapp.com/api/v1/units',
      technologies: 'https://age-of-empires-2-api.herokuapp.com/api/v1/technologies',
      civilizations: 'https://age-of-empires-2-api.herokuapp.com/api/v1/civilizations',
    },
    location: false,
  }

  let wrapper

  beforeEach(() => {
    wrapper = shallow(
      <Provider store={store}>
        <Structures {...props} />
      </Provider>
    )
  })

  it('should match snapshot', () => {
    expect(wrapper).toMatchSnapshot()
  })

  it('should render 1 <Structures /> component', () => {
    expect(wrapper).toHaveLength(1)
  })
})
