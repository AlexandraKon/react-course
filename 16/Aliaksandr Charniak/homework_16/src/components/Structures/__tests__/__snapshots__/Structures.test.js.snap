// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Structures /> mounting behaviour should match snapshot 1`] = `
<ContextProvider
  value={
    Object {
      "store": Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      },
      "subscription": Subscription {
        "handleChangeWrapper": [Function],
        "listeners": Object {
          "notify": [Function],
        },
        "onStateChange": [Function],
        "parentSub": undefined,
        "store": Object {
          "clearActions": [Function],
          "dispatch": [Function],
          "getActions": [Function],
          "getState": [Function],
          "replaceReducer": [Function],
          "subscribe": [Function],
        },
        "unsubscribe": null,
      },
    }
  }
>
  <Connect(Structures)
    data={
      Object {
        "civilizations": "https://age-of-empires-2-api.herokuapp.com/api/v1/civilizations",
        "structures": Array [
          Object {
            "age": "Dark",
            "armor": "0/7",
            "build_time": 50,
            "cost": Object {
              "Wood": 175,
            },
            "expansion": "Age of Kings",
            "hit_points": 1200,
            "id": 1,
            "line_of_sight": 5,
            "name": "Barracks",
            "special": Array [
              "Garrison: 10 created units",
            ],
          },
          Object {
            "age": "Dark",
            "armor": "0/7",
            "build_time": 35,
            "cost": Object {
              "Wood": 150,
            },
            "expansion": "Age of Kings",
            "hit_points": 1800,
            "id": 2,
            "line_of_sight": 5,
            "name": "Dock",
            "special": Array [
              "Garrison: 10 created units",
            ],
          },
        ],
        "technologies": "https://age-of-empires-2-api.herokuapp.com/api/v1/technologies",
        "units": "https://age-of-empires-2-api.herokuapp.com/api/v1/units",
      }
    }
    location={false}
  />
</ContextProvider>
`;
